### ISSUE TITLE
[Feature] Pre-flight Carbon Estimation — Estimate Emissions Before Training Starts

### ISSUE THREAD

**Issue by Tamish03** [mlco2/codecarbon#1174]
Summary
CodeCarbon currently measures carbon emissions after a training run completes. This is too late for decision-making. A researcher choosing between BERT-base and BERT-large fine-tuning, or between running 50 vs 100 epochs, has no way to understand the carbon tradeoff before committing compute. This proposal introduces a CarbonEstimator module that produces pre-flight emission estimates from model architecture metadata, training configuration, hardware profile, and location.

Problem
Two real scenarios this blocks today:
Scenario A — Architecture selection: A team is deciding between three candidate architectures for a production model. They run all three to compare accuracy. Only after all three complete do they discover that the winning architecture used 4× more carbon than the second-best while delivering only 1.2% better accuracy. With pre-flight estimation, this tradeoff is visible before committing.
Scenario B — HPO scope: A researcher plans a 200-trial hyperparameter sweep. With estimation, they could see upfront: "this sweep will cost approximately 8.4 kg CO2 on your hardware in your region" — and perhaps scope it to 80 trials instead.

Proposed Solution
A CarbonEstimator class in codecarbon/estimation/ that computes emission estimates through three methods:
Method 1 — Linear estimator (always available, no ML):
estimated_emissions = (flops / peak_hardware_flops) × hardware_tdp × duration_hours × carbon_intensity
Where FLOP count is estimated from model parameter count and input shape, duration from dataset size / batch size / epochs.
Method 2 — Regression estimator (optional, requires scikit-learn):
Trained on CodeCarbon's own aggregated emission dataset. Inputs: (model_category, param_count, dataset_size, batch_size, epochs, hardware_class, region). Outputs: point estimate + 90% confidence interval.
Method 3 — Hardware-corrected estimator:
Uses actual measured TDP from CodeCarbon's existing hardware database, adjusted for observed utilization patterns from historical runs on the same hardware.

API Design
pythonfrom codecarbon.estimation import CarbonEstimator

estimator = CarbonEstimator()

# From a PyTorch model object
result = estimator.estimate(
    model=my_pytorch_model,
    dataset_size=50_000,
    batch_size=32,
    epochs=10,
    hardware="A100",        # or auto-detect
    region="france"         # or auto-detect from IP
)

print(result.point_estimate_kg)   # 0.42
print(result.confidence_interval) # (0.31, 0.58)
print(result.breakdown)
# {gpu: 78%, cpu: 14%, ram: 8%}

print(result.equivalences)
# "Equivalent to 3.2 km driven by an average car"

# Comparison mode
result = estimator.compare(
    models=[bert_base, bert_large, roberta_base],
    dataset_size=100_000,
    epochs=5,
    region="germany"
)
result.plot_comparison()  # carbon vs param_count scatter
CLI usage:
bashcodecarbon estimate \
  --model-params 110M \
  --model-type transformer \
  --dataset-size 50000 \
  --epochs 10 \
  --hardware A100 \
  --region france

Architecture
codecarbon/estimation/
    ├── __init__.py
    ├── estimator.py          # CarbonEstimator main class
    ├── flop_counter.py       # FLOP estimation from model metadata
    ├── hardware_model.py     # TDP lookup + utilization modeling
    ├── regression_model.py   # sklearn-based estimator (optional)
    └── result.py             # EstimationResult dataclass

What This Does NOT Touch

Core tracker — estimation is entirely separate from measurement
Existing CSV schema
Existing public API
No new required dependencies (regression model requires scikit-learn but gracefully degrades to linear estimator)

Open Questions for Maintainers

Where should the pre-trained regression model artifact live? Options: (a) bundled in the package, (b) downloaded on first use from a hosted URL, (c) users train their own from their accumulated data.
Should FLOPs counting support TensorFlow and JAX models, or start with PyTorch only?
Is there an existing dataset of CodeCarbon emissions data (aggregated, anonymized) that could be used to train the regression model?
Should estimation be accessible as a standalone CLI subcommand (codecarbon estimate) or only via the Python API?

Implementation Plan

 PR 1: EstimationResult dataclass + linear estimator + hardware TDP lookup
 PR 2: FLOP counter supporting PyTorch models via parameter introspection
 PR 3: CLI subcommand codecarbon estimate
 PR 4: Regression estimator with scikit-learn (once dataset question is resolved)
 PR 5: Comparison mode + visualization

I'd like to start with PR 1 once maintainers confirm the module structure. Happy to adjust architecture based on feedback.
